### PR TITLE
Add EntityDynamicType.removeEffector(Effector)

### PR DESCRIPTION
**Pull request by sjcorbett**
Not sure if this is controversial.

### CODE REVIEW

**Comment by alasdairhodge** [comment]
Awesome. Not controversial, IMO; the symmetry makes perfect sense.

**Comment by alasdairhodge** [comment]
Controversial lack of a unit test, however.

**Comment by sjcorbett** [comment]
@alasdairhodge Thanks for comments. Would you cast your eye over a second time? Note the comment on `removeEffector` and the assertions of the new test. Brooklyn will still invoke removed effectors because it knows better than you and gets Groovy to look up the method dynamically.

File: core/src/main/java/brooklyn/entity/basic/EntityDynamicType.java [hunk 1]
```diff
@@ -130,17 +130,38 @@ public void addEffector(Effector<?> newEffector) {
...
         addEffector(effector, new EffectorBodyTaskFactory<T>(body));
     }
 
+    /**
+     * Removes the given {@link Effector} from this entity.
+     * <p>
+     * Note that if the argument is an instance of {@link EffectorWithBody} it will
+     * still be possible to invoke the effector on the entity by calling
+     * <code>entity.invoke(effector, argumentsMap)</code>.
+     */
```
Comment (alasdairhodge) [inline core/src/main/java/brooklyn/entity/basic/EntityDynamicType.java]: The ability to invoke removed efectors is a bug, this comment should reflect that. Is it trivial to fix `entity.invoke()`?

**Comment by alasdairhodge** [comment]
Unfortunate behaviour in `entity.invoke()`; I'd call that a bug, and suggest a fix in the form of an explicit check in `Effectors.invocation()`, which already throws `UnsupportedOperationException` for missing effectors.

**Comment by sjcorbett** [comment]
Is it a bug? We can remove the effector but for example `Startable` entities are still going to have a `stop` method the can be invoked programatically. Brooklyn turns those method calls into effector invocations. Should we not do this if the related effector has been removed?

I'm wary of modifying `Effectors.invocation` (see https://github.com/apache/incubator-brooklyn/blob/master/core/src/main/java/brooklyn/entity/effector/Effectors.java#L153-159) without looking more into what is relying on this at the moment.

**Comment by neykov** [comment]
I don't see it as a bug. Effectors are like enrichers - the entity might not be aware of it at all. The entity could override an effector body by adding it to the `entityType`, but otherwise the caller's implementation will be used.
Take for example an `SshEffector` which executes a shell command on the entity's machine. It can be executed on any entity which uses `SshMachineLocation`, but it's not feasible to add it to all such entities.

**Comment by alasdairhodge** [comment]
Good points, @sjcorbett and @neykov, thanks. I agree that "fixing" `Effectors` is too risky to consider here. I would still consider it a bug if effectors that have been *explicitly removed in code* can still be invoked.

**Comment by alasdairhodge** [comment]
Interesting example, @svet. Just out of curiosity: this specific usage feels like a band-aid for lacking the right place to define a (static) `executeCommand` effector. Would you dynamically add the mixin effector if an explicit `SshMachine` entity was present in the mgmt tree? (Not that I'm wanting to reopen the "locations as entities" discussion...)

**Comment by neykov** [comment]
Well I wouldn't add it dynamically at all, just invoke it on the entity (not sure if I got your question right). 
There are other examples of cross-cutting concerns which would work on a large number of entities, but none would add them explicitly to the "favoured" list. Like an effector which reports for potential rebind problems on running entities.

**Comment by alasdairhodge** [comment]
@neykov: Hmm, interesting and powerful. The (undeclared) "cross-cutting concern" effectors won't be shown in the UI, right? And can't be invoked via REST?

**Comment by alasdairhodge** [comment]
@sjcorbett: this thread has been a real eye-opener. I'm coming round to the elegance of "unattached" effectors for cross-cutting concerns, but this does call into question the notion of removing effectors, which is what this PR is all about.

In our specific case, can we instead prevent the unwanted effector being added in the first place?

**Comment by neykov** [comment]
>  The (undeclared) "cross-cutting concern" effectors won't be shown in the UI, right? And can't be invoked via REST?

That's right. We could add functionality to invoke undeclared effectors to the REST API but it's not supported now.

On the other hand the effector task will appear in the entity's Activity tab for inspection.

**Comment by neykov** [comment]
Removing effectors could still be useful for UI reasons.

**Comment by aledsage** [comment]
Thanks all - merging now.

**Pull request merged**